### PR TITLE
provide individual gateway.enabled flags for addons

### PR DESCRIPTION
**Pull request by jmazzitelli**
I took the work that @cezarsa did for https://github.com/istio/istio/pull/9640 and fixed it up so it has no conflicts, is based on release-1.1 branch, and I added kiali to it. (we can close that PR out, this supercedes it).

This does NOT support https yet. I am working on that. But I want to get this in now and base the https work on this.

@sdake @douglas-reid please review. I ran this with the following settings, and I was able to go through the gateway to get to things like jaeger, grafana, prometheus:

```
--set grafana.gateway.enabled=true
--set jaeger.gateway.enabled=true
--set kiali.gateway.enabled=true
--set prometheus.gateway.enabled=true
```

Once merged, I can sprinkle magic fairy dust on this to get https to work, once I find out where I can buy some :)

### CODE REVIEW

**Comment by douglas-reid** [comment]
Is this missing the jaeger edits?

**Comment by objectiser** [comment]
@jmazzitelli @douglas-reid Do we need to add another flag for the gateways? Previously they were enabled if the add-on was enabled - isn't this enough?

For tracing - there is an existing gateway - but it currently has no flag. So that needs to be added - either based on `tracing.enabled`, `tracing.gateway.enabled` or possibly both? (i.e. it should not be possible to enable a gateway for an add-on, if the add-on itself is not enabled).

File: install/kubernetes/helm/subcharts/grafana/templates/gateway.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{{- if .Values.gateway.enabled }}
```
Comment (objectiser) [inline install/kubernetes/helm/subcharts/grafana/templates/gateway.yaml]: As mentioned in main comment, an add-on gateway should not be enabled if the add-on itself is not enabled - so needs `.Values.enabled and .Values.gateway.enabled`.

Comment (jmazzitelli) [reply install/kubernetes/helm/subcharts/grafana/templates/gateway.yaml]: I fixed this - all the conditionals now "and" both the enabled and gateway.enabled flags:

```
{{- if and (.Values.enabled) (.Values.gateway.enabled) }}
```

**Comment by jmazzitelli** [comment]
> For tracing - there is an existing gateway - but it currently has no flag. So that needs to be added

This is also done in the latest commit.

File: install/kubernetes/helm/subcharts/tracing/values.yaml [hunk 2]
```diff
@@ -66,3 +66,6 @@ ingress:
     #   hosts:
     #     - tracing.local
 gatewayName: ingressgateway
```
Comment (objectiser) [inline install/kubernetes/helm/subcharts/tracing/values.yaml]: Would it be possible to move this under the gatrway node, so:

```
gateway:
  enabled: false
  name: ingressgateway
```
I

**Comment by jmazzitelli** [comment]
I just tested my with my latest commit. without gateway.enabled=true (but with enabled=true) all add-ons are not exposed ... witih gateway.enabled=true, they are exposed. So that "and" conditional across all the gateway.yaml files are working.

File: install/kubernetes/helm/subcharts/tracing/values.yaml (hunk 2 above)
Comment (jmazzitelli) [reply install/kubernetes/helm/subcharts/tracing/values.yaml]: Yup - done. See commit 822c690a51dc3eb1afdd7ac61eedaa162dcbfca2

**Comment by costinm** [comment]
I think a better approach ( in context of the new Gateway API, and the
'a-la-carte' discussions ) is to move all this into the gateway
template/config.
It depends on implementing parts of the new API - but it has a lot of
benefits, including easier to set TLS.

In the new model the Gateway is defined in the namespace of the ingress (
where certs are also stored ), and delegates domains to specific namespaces.
Certmanager worked in 1.0 - would be great to test it again, but it is also
possible to use a wildcard certificate. ACME requires DNS verification for
wildcards,
I got one using 'lego', but you need one of the supported providers. I
would mention Citadel has a flag to generate DNS-based certs as well - and
it is possible
to import Citadel root into a browser for testing.

For authentication - it would be great to use Istio APIs, we support JWTs
and RBAC. I did some experiments with MTLS auth - you can convert
an Istio cert into a PFX file and import it in browser - it is a bit
inconvenient but appears to work. For some large users provisioning the
admin browsers with
an enterprise CA and certs is not bad.

**Comment by jmazzitelli** [comment]
the prometheus service account does not follow the convention for all the others (all others have service account names of "ABC-service-account" such as "grafana-service-account") but prometheus just has "prometheus" as the service account name. I'm going to create another commit to this PR so the prometheus SA is named "prometheus-service-account" to ensure it is following the conventions for the rest of the project.

File: install/kubernetes/helm/subcharts/prometheus/templates/serviceaccount.yaml [hunk 3]
```diff
@@ -7,7 +7,7 @@ imagePullSecrets:
 {{- end }}
 {{- end }}
 metadata:
-  name: prometheus
+  name: prometheus-service-account
```
Comment (sdake) [inline install/kubernetes/helm/subcharts/prometheus/templates/serviceaccount.yaml]: is there upgrade impact with this change?  Can this change be broken out separately so we can revert it if needed when we go through upgrade testing prior to 1.1 release?

File: install/kubernetes/helm/subcharts/tracing/values.yaml [hunk 4]
```diff
@@ -65,4 +65,7 @@ ingress:
     # - secretName: tracing-tls
     #   hosts:
     #     - tracing.local
-gatewayName: ingressgateway
+
+gateway:
+  enabled: false
+  name: ingressgateway
```
Comment (sdake) [inline install/kubernetes/helm/subcharts/tracing/values.yaml]: why does Jaeger permit setting the ingress, while other addons don't?  Probably something to handle in a followon PR.

File: install/kubernetes/helm/subcharts/prometheus/templates/serviceaccount.yaml (hunk 3 above)
Comment (jmazzitelli) [reply install/kubernetes/helm/subcharts/prometheus/templates/serviceaccount.yaml]: I removed that commit from this PR and created a new PR with just that change: https://github.com/istio/istio/pull/10362

File: install/kubernetes/helm/subcharts/tracing/values.yaml (hunk 4 above)
Comment (jmazzitelli) [reply install/kubernetes/helm/subcharts/tracing/values.yaml]: I do not know - I wasn't the author of that :) This just moves the existing setting under the "gateway" section, as suggested by @objectiser.

Comment (sdake) [reply install/kubernetes/helm/subcharts/tracing/values.yaml]: sounds fine.  thanks.

File: install/kubernetes/helm/subcharts/prometheus/templates/serviceaccount.yaml (hunk 3 above)
Comment (sdake) [reply install/kubernetes/helm/subcharts/prometheus/templates/serviceaccount.yaml]: thank you.

**Comment by sdake** [comment]
@costinm I'm good to merge this as is assuming you agree to an incremental approach to getting to https://github.com/istio/istio/pull/10349#issuecomment-445468816 (hopefully in 1.1).  I am more concerned at this point about having HTTPS endpoints for addons and this is the first step to that work.

Cheers
-steve

**Comment by jmazzitelli** [comment]
follow-on PR that puts https in front of the add-ons (prometheus, grafana, jaeger, kiali): https://github.com/istio/istio/pull/10363

**Comment by sdake** [comment]
@jmazzitelli can you amend the first commit message in this PR and add @cezarsa as a co-author?  `Co-Authored-By name <email>`.

Cheers
-steve

**Comment by jmazzitelli** [comment]
k.. as per [github docs](https://help.github.com/articles/creating-a-commit-with-multiple-authors/), I'll add
`Co-authored-by: Cezar Sá Espinola <cezarsa@gmail.com>`

**Comment by costinm** [comment]
Sorry, can we wait few more days ? If new Gateway proposal is accepted, I
would rather start using it and move all ingress to the new style.

**Comment by jmazzitelli** [comment]
> @jmazzitelli my apologies if this work ends up needing a refactor

No worries - I just want these add-ons exposed via gateway/https after an install - however we get it done, we get it done :-)

**Comment by costinm** [comment]
@sdake - yes, that's the document, and the main change is that the Gateway is no longer spread to each virtual service, instead it is co-located with the actual workload.

For internal services and 1.1 - all are in Istio system, but we still want to group them all, so we can start moving telemetry and other things to separate namespaces ( for users who don't want the admins of grafana to have access to Citadel master CA secret for example ).

I guess we can just merge it as is and refactor later. 

Is there any way to move the values.yaml settings under ingress or global - instead of under individual components ? Maybe with some way to customize how they are mapped too - right now the templates use hardcoded ports, we may want to use proper domain names ( grafana.prod.example.com ).

**Comment by jmazzitelli** [comment]
@sdake what do you think? Merge this now? We need the cla label put back on to kickstart the workflow again.

**Comment by sdake** [comment]
@costinm 
re hardcoded ports, I agree - this is not ideal.  I think what you are proposing is documented here:
https://preliminary.istio.io/docs/tasks/traffic-management/secure-ingress/#configure-a-tls-ingress-gateway-for-multiple-hosts

Implementing this as a follow-on would be really cool and hopefully finish the job on addon exports over one IP address with multiple DNS cnames.  (At least in my testing on bare metal, this worked propelry in my personal lab in the 1.0 release).

@jmazzitelli one of the google peeps will have to add the CLA label.  Otherwise the PR:

Thanks for tackling this very important subject (IPs are very expensive in clouds, and LBs require IP addresses).

/lgtm
/approve

**Comment by jmazzitelli** [comment]
> @jmazzitelli one of the google peeps will have to add the CLA label.

@douglas-reid we need your google super-powers to kick this PR so it can be merged.

**Comment by jmazzitelli** [comment]
@GregHanson  does this need to be "cla: yes" for the merge to happen? Because I still see the "cla/google" check not passing fully "CLAs are signed, but unable to verify author consent". I don't know what it is looking for in order to "verify author consent".

**Comment by jmazzitelli** [comment]
@cezarsa can you post a comment in here saying you consent? I see you added a emojii - but it might be required that you post an actual comment with your consent.

**Comment by cezarsa** [comment]
Ops, sorry, didn't know that. I consent! :)

**Pull request merged**